File: server/favicon.test.ts
import { test, before, after, beforeEach } from 'node:test';
import assert from 'node:assert/strict';
import { createServer } from 'node:http';
import type { Server } from 'node:http';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { unlinkSync } from 'node:fs';

// Isolate every test run in its own temp DB so tests are repeatable.
const TEST_DB_PATH = join(tmpdir(), `rss-favicon-test-${process.pid}.db`);
process.env.TEST_DB = TEST_DB_PATH;

const { app } = await import('./app.ts');
const { db } = await import('./db.ts');
const { getFavicon, DEFAULT_FAVICON } = await import('./favicon.ts');

// ── Helpers ───────────────────────────────────────────────────────────────────

const realFetch = globalThis.fetch;
let server: Server;
let baseUrl: string;

/** Replace global fetch with a counter-tracking stub returning the given bytes. */
function stubFetch(bytes: Buffer | null, contentType = 'image/png') {
  const calls = { n: 0 };
  globalThis.fetch = (async () => {
    calls.n++;
    if (bytes === null) throw new Error('network down');
    return new Response(bytes as unknown as BodyInit, { status: 200, headers: { 'content-type': contentType } });
  }) as typeof fetch;
  return calls;
}

before(async () => {
  await new Promise<void>(resolve => {
    server = createServer(app);
    server.listen(0, '127.0.0.1', () => {
      baseUrl = `http://127.0.0.1:${(server.address() as { port: number }).port}`;
      resolve();
    });
  });
});

// Test-client requests must use the real fetch; only the route's *upstream* call
// is stubbed (both share globalThis.fetch in-process).
const httpGet = (path: string) => realFetch(`${baseUrl}${path}`);

beforeEach(() => {
  globalThis.fetch = realFetch; // reset any stub leaked from a prior test
  db.prepare('DELETE FROM favicon_cache').run();
});

after(async () => {
  globalThis.fetch = realFetch;
  await new Promise<void>(resolve => server.close(() => resolve()));
  db.close();
  try { unlinkSync(TEST_DB_PATH); } catch {}
});

const PNG = Buffer.from([0x89, 0x50, 0x4e, 0x47]); // PNG magic bytes

// ── getFavicon (module contract) ──────────────────────────────────────────────

test('getFavicon — returns null for an invalid domain (no fetch, no row)', async () => {
  const calls = stubFetch(PNG);
  assert.equal(await getFavicon('not a domain'), null);
  assert.equal(await getFavicon('anthropic'), null);   // single-label hostname
  assert.equal(calls.n, 0, 'invalid input must not hit the network');
  assert.equal((db.prepare('SELECT COUNT(*) n FROM favicon_cache').get() as { n: number }).n, 0);
});

test('getFavicon — miss fetches upstream and persists the bytes', async () => {
  const calls = stubFetch(PNG, 'image/x-icon');
  const result = await getFavicon('example.com');
  assert.equal(calls.n, 1);
  assert.deepEqual(result!.image, PNG);
  assert.equal(result!.contentType, 'image/x-icon');

  const row = db.prepare('SELECT image, content_type FROM favicon_cache WHERE domain = ?').get('example.com') as { image: Buffer; content_type: string };
  assert.deepEqual(row.image, PNG);
});

test('getFavicon — second call is served from cache without re-fetching', async () => {
  const calls = stubFetch(PNG);
  await getFavicon('example.com');
  await getFavicon('example.com');
  assert.equal(calls.n, 1, 'upstream should be hit only once');
});

test('getFavicon — upstream failure returns null and negative-caches', async () => {
  const calls = stubFetch(null);
  assert.equal(await getFavicon('broken.com'), null);
  assert.equal(calls.n, 1);

  const row = db.prepare('SELECT image FROM favicon_cache WHERE domain = ?').get('broken.com') as { image: Buffer | null };
  assert.equal(row.image, null, 'negative cache stores a NULL image');

  // A fresh negative row should be served without another upstream hit.
  assert.equal(await getFavicon('broken.com'), null);
  assert.equal(calls.n, 1, 'negative cache should suppress the retry within TTL');
});

test('getFavicon — empty upstream body is treated as a failure', async () => {
  stubFetch(Buffer.alloc(0));
  assert.equal(await getFavicon('empty.com'), null);
});

// ── GET /api/favicon (route behavior) ─────────────────────────────────────────

test('GET /api/favicon — serves the real icon with a 7-day cache header', async () => {
  stubFetch(PNG, 'image/png');
  const res = await httpGet(`/api/favicon?domain=example.com`);
  assert.equal(res.status, 200);
  assert.match(res.headers.get('cache-control') ?? '', /max-age=604800/);
  assert.deepEqual(Buffer.from(await res.arrayBuffer()), PNG);
});

test('GET /api/favicon — invalid domain returns the default image (200, not an error)', async () => {
  const res = await httpGet(`/api/favicon?domain=anthropic`);
  assert.equal(res.status, 200);
  assert.match(res.headers.get('content-type') ?? '', /image\/svg\+xml/);
  assert.deepEqual(Buffer.from(await res.arrayBuffer()), DEFAULT_FAVICON);
});

test('GET /api/favicon — upstream failure returns the default image (200)', async () => {
  stubFetch(null);
  const res = await httpGet(`/api/favicon?domain=broken.com`);
  assert.equal(res.status, 200);
  assert.match(res.headers.get('content-type') ?? '', /image\/svg\+xml/);
  assert.match(res.headers.get('cache-control') ?? '', /max-age=86400/);
  assert.deepEqual(Buffer.from(await res.arrayBuffer()), DEFAULT_FAVICON);
});

test('GET /api/favicon — missing domain returns the default image (200)', async () => {
  const res = await httpGet(`/api/favicon`);
  assert.equal(res.status, 200);
  assert.deepEqual(Buffer.from(await res.arrayBuffer()), DEFAULT_FAVICON);
});

File: server-go/internal/favicon/favicon_test.go
package favicon_test

import (
	"context"
	"errors"
	"sync/atomic"
	"testing"

	"rss-reader/server-go/internal/db"
	"rss-reader/server-go/internal/favicon"
)

func newTestDB(t *testing.T) *db.DB {
	t.Helper()
	handle, err := db.OpenHandle(t.TempDir() + "/t.db")
	if err != nil {
		t.Fatalf("open: %v", err)
	}
	if err := db.InitSchema(handle.Writer()); err != nil {
		t.Fatalf("schema: %v", err)
	}
	t.Cleanup(func() { handle.Close() })
	return handle
}

func TestInvalidDomainNoFetch(t *testing.T) {
	var calls int32
	c := favicon.New(newTestDB(t), func(context.Context, string) ([]byte, string, error) {
		atomic.AddInt32(&calls, 1)
		return nil, "", nil
	})
	for _, d := range []string{"", "not a domain", "http://x.com", "localhost", "a..b.com"} {
		res, err := c.Get(context.Background(), d)
		if err != nil || res != nil {
			t.Errorf("invalid domain %q: res=%v err=%v (want nil,nil)", d, res, err)
		}
	}
	if calls != 0 {
		t.Errorf("invalid domains triggered %d fetches, want 0", calls)
	}
}

func TestCachesBlobAndServesFromCache(t *testing.T) {
	var calls int32
	c := favicon.New(newTestDB(t), func(context.Context, string) ([]byte, string, error) {
		atomic.AddInt32(&calls, 1)
		return []byte("PNGDATA"), "image/png", nil
	})
	// First call fetches + caches.
	res, err := c.Get(context.Background(), "example.com")
	if err != nil || res == nil || string(res.Image) != "PNGDATA" || res.ContentType != "image/png" {
		t.Fatalf("first get: res=%v err=%v", res, err)
	}
	// Second call served from cache (no new fetch).
	res, err = c.Get(context.Background(), "example.com")
	if err != nil || res == nil || string(res.Image) != "PNGDATA" {
		t.Fatalf("cached get: res=%v err=%v", res, err)
	}
	if calls != 1 {
		t.Errorf("expected 1 fetch (rest cached), got %d", calls)
	}
}

func TestNegativeCacheAndTTLRetry(t *testing.T) {
	var calls int32
	failing := func(context.Context, string) ([]byte, string, error) {
		atomic.AddInt32(&calls, 1)
		return nil, "", errors.New("upstream down")
	}
	handle := newTestDB(t)
	c := favicon.New(handle, failing)
	const t0 int64 = 1_700_000_000_000
	now := t0
	favicon.SetClock(c, func() int64 { return now })

	// Failure → nil result, negative cache written.
	if res, err := c.Get(context.Background(), "fail.example.com"); err != nil || res != nil {
		t.Fatalf("failing get: res=%v err=%v", res, err)
	}
	// Within negative TTL (< 1 day): no retry.
	now = t0 + 60*1000
	if res, _ := c.Get(context.Background(), "fail.example.com"); res != nil {
		t.Fatal("expected nil within negative TTL")
	}
	if calls != 1 {
		t.Fatalf("negative cache not honored: %d fetches, want 1", calls)
	}
	// After negative TTL (> 1 day): retry fires.
	now = t0 + 25*60*60*1000
	_, _ = c.Get(context.Background(), "fail.example.com")
	if calls != 2 {
		t.Fatalf("expected retry after negative TTL: %d fetches, want 2", calls)
	}
}

func TestEmptyBodyIsNegative(t *testing.T) {
	var calls int32
	c := favicon.New(newTestDB(t), func(context.Context, string) ([]byte, string, error) {
		atomic.AddInt32(&calls, 1)
		return []byte{}, "image/png", nil // empty body → treat as failure
	})
	if res, err := c.Get(context.Background(), "empty.example.com"); err != nil || res != nil {
		t.Fatalf("empty body: res=%v err=%v (want nil,nil)", res, err)
	}
	if calls != 1 {
		t.Errorf("expected 1 fetch, got %d", calls)
	}
}

func TestDefaultFaviconIsSVG(t *testing.T) {
	if favicon.DefaultContentType != "image/svg+xml" {
		t.Errorf("default content type: %q", favicon.DefaultContentType)
	}
	if len(favicon.DefaultFavicon) == 0 {
		t.Error("default favicon empty")
	}
}
